### PR TITLE
feat: support arbitrary message signing

### PR DESCRIPTION
**Pull request by beguene**
This adds a doSignature helper and other methods to request a signature 
For more context see
https://github.com/hirosystems/stacks-wallet-web/issues/1051
and 
https://github.com/hirosystems/stacks-wallet-web/pull/2350

### CODE REVIEW

File: packages/connect-react/src/react/hooks/use-connect.ts [hunk 1]
```diff
@@ -106,6 +108,14 @@ export const useConnect = () => {
     });
   }
 
+  function doSignature(options: SignatureRequestOptions) {
```
Comment (kyranjamie) [inline packages/connect-react/src/react/hooks/use-connect.ts:111]: I'd recommend we move away from this `do` prefixed messaging. I know it's already used here in some places, but hopefully we can kill that too eventually (it's bad english and I have no idea why we ever used it 🙈 )

Suggest: `createSignature`, `makeSignature` etc

File: packages/connect/src/signature/index.ts [hunk 2]
```diff
@@ -0,0 +1,93 @@
...
+  const stxAddresses = userSession?.loadUserData().profile?.stxAddress;
+  const chainIdToKey = {
+    [ChainID.Mainnet]: 'mainnet',
+    [ChainID.Testnet]: 'testnet',
+  };
+  const address: string | undefined = stxAddresses?.[chainIdToKey[network.chainId]];
+  return address;
+}
+
+function getDefaults(options: SignatureRequestOptions) {
```
Comment (kyranjamie) [inline packages/connect/src/signature/index.ts:27]: Suggest more descriptive name

File: packages/connect/src/signature/index.ts [hunk 3]
```diff
@@ -0,0 +1,93 @@
...
+  makeTokenFn: (options: T) => Promise<string>
+) {
+  const token = await makeTokenFn({
+    ...getDefaults(options),
+    ...options,
+  } as T);
+  return openSignaturePopup({ token, options });
+}
+
+export function openSignature(options: SignatureRequestOptions) {
```
Comment (kyranjamie) [inline packages/connect/src/signature/index.ts:91]: Suggest more descriptive name here, too. `openSignatureRequestPopup`?

File: packages/connect/src/types/signature.ts [hunk 4]
```diff
@@ -0,0 +1,45 @@
...
+   * This is set by default if a `userSession` option is provided.
+   */
+  stxAddress?: string;
+  appDetails?: AuthOptions['appDetails'];
+  network?: StacksNetwork;
+}
+
+export interface SignatureData {
+  signature: string; // - Hex encoded DER signature
+  publicKey: string; // - Hex encoded private string taken from privateKey
```
Comment (kyranjamie) [inline packages/connect/src/types/signature.ts:44]: I generally prefer `//` comments, but editors will show this info inline if you use `/**` comments above the line

File: packages/connect/src/signature/index.ts [hunk 5]
```diff
@@ -0,0 +1,93 @@
...
+    network,
+    userSession,
+  };
+  return {
+    stxAddress: getStxAddress(defaults),
+    ...defaults,
+  };
+}
+
+const signPayload = async (payload: SignatureRequestPayload, privateKey: string) => {
```
Comment (kyranjamie) [inline packages/connect/src/signature/index.ts:41]: Can use `async function` for consistency with rest of file

Comment (aulneau) [reply packages/connect/src/signature/index.ts:41]: the type `SignatureRequestPayload` doesn't match what is being passed to it here: https://github.com/hirosystems/connect/pull/205/files#diff-802b39cd7bbe97b86e253cf1302d332785dff1104174eeee067620aeb806cc41R72, it's possible that many more options are being passed to this `signPayload` method.

File: packages/connect/src/types/signature.ts [hunk 6]
```diff
@@ -0,0 +1,47 @@
...
+export interface SignatureOptions {
+  message: string;
+  onFinish?: Finished;
+  onCancel?: Canceled;
+}
+
+export type SignaturePopup = {
+  token: string;
+  options: SignatureOptions;
+};
```
Comment (kyranjamie) [inline packages/connect/src/types/signature.ts:28]: This should just be an interface

**Review by kyranjamie — APPROVED** [review]

**Review by fbwoolf — APPROVED** [review]

**Pull request merged**